bugsnag-android-core/src/main/java/com/bugsnag/android/Client.java

```diff
@@ -172,7 +172,7 @@ public Unit invoke(Boolean hasConnection, String networkState) {
         Resources resources = appContext.getResources();
         deviceDataCollector = new DeviceDataCollector(connectivity, appContext,
                 resources, deviceId, info, Environment.getDataDirectory(),
-                new RootDetector(), logger);
+                new RootDetector(logger), logger);
```

Comment:
I'd be against instantiating `RootDetector` within `DeviceDataCollector` because passing the object in via the constructor instead increases the testability of the class through allowing dependency injection. Reusing the logger could also couple `RootDetector` and `DeviceDataCollector` which doesn't seem ideal.

An example of why this is helpful is in this test, where a [mocked `RootDetector` class](https://github.com/bugsnag/bugsnag-android/blob/setup-native-library/bugsnag-android-core/src/test/java/com/bugsnag/android/DeviceDataCollectorSerializationTest.kt#L30) is passed in. This makes it easy to assert against the behaviour of `DeviceDataCollector` when `isRooted()` returns true or false, which would otherwise be hard to control.

I definitely agree there are a lot of parameters - this is a common problem in the codebase and there's a ticket to investigate more manageable approaches to dependency injection. Ultimately I think it'd make sense to [create modules](https://developer.android.com/training/dependency-injection/hilt-android#inject-provides) (or something similar) that define the dependency graph in a declarative way.

Reply:
You're quite right, DI is a compelling reason for doing it that way.  It just didn't occur to me for some reason.